src/Microsoft.AspNetCore.Rewrite/UrlRewrite/InputParser.cs

```diff
@@ -0,0 +1,212 @@
+                }
+                else if (context.Current == CloseBrace)
+                {
+                    // TODO we should be throwing a syntax error if we have uneven close braces
+                    // Can fix by keeping track of the number of '{' and '}' with an int, where {
+                    // increments and } decrements. Throw if < 0.
+                    return new Pattern(results);
+                }
+                else
+                {
+                    // Parse for literals, which will return on either the end of the test string 
+                    // or when it hits a special character
+                    ParseLiteral(context, results);
+                }
+            }
+            return new Pattern(results);
+        }
+
+        private static void ParseParameter(ParserContext context, List<PatternSegment> results)
+        {
+            context.Mark();
+            // Four main cases:
+            // 1. {NAME} - Server Variable, create lambda to get the part of the context
+            // 2. {R:1}  - Rule parameter
+            // 3. {C:1}  - Condition Parameter
+            // 4. {function:xxx} - String function 
+            // TODO consider perf here. This is on startup and will only happen one time
+            // (unless we support Reload)
+            string parameter;
+            while (context.Next())
+            {
+                if (context.Current == CloseBrace)
+                {
+                    // This is just a server variable, so we do a lookup and verify the server variable exists.
+                    parameter = context.Capture();
+                    results.Add(new PatternSegment(ServerVariables.FindServerVariable(parameter)));
+                    return;
+                }
+                else if (context.Current == Colon)
+                {
+                    parameter = context.Capture();
+                    
+                    // Only 5 strings to expect here. Case sensitive.
+                    switch (parameter)
+                    {
+                        case "ToLower":
+                            {
+                                var pattern = ParseString(context);
+                                results.Add(new PatternSegment((ctx, ruleMatch, condMatch) =>
+                                {
+                                    var str = pattern.Evaluate(ctx, ruleMatch, condMatch);
+                                    return str.ToLowerInvariant();
+                                }));
+
+                                // at this point, we expect our context to be on the ending closing brace,
+                                // because the ParseString() call will increment the context until it 
+                                // has processed the new string.
+                                if (context.Current != CloseBrace)
+                                {
+                                    throw new FormatException();
```

Comment:
The comment above should be the exception's message. Otherwise, I get a `FormatException` but I don't know why/where. The exception should also include the character number